Crash when a test function is decorated with ``@pytest.fixture``

## Question
I am using pylint 2.8.3 and one of my repos pipeline fails for the following exception within pylint:

```
i is not None and i.qname() in qnames or i.name in qnames

 TypeError: 'in <string>' requires string as left operand, not Uninferable
```

I cannot work out what is causing this but if I remove a failing module then this exception just gets thrown in another module.

Can anyone assist me with debugging this issue?

Here is the stack trace in the build logs:

```
Exception on node <Call l.92 at 0x7f9cb66bc220> in file '/codebuild/output/src662162499/src/github.com/edfenergy/ind-openit-export-forecasting-service/src/test/unit/export_forecasting_api_tests.py'
--
462 | Traceback (most recent call last):
463 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/bin/pylint", line 8, in <module>
464 | sys.exit(run_pylint())
465 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/__init__.py", line 24, in run_pylint
466 | PylintRun(sys.argv[1:])
467 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/run.py", line 384, in __init__
468 | linter.check(args)
469 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 962, in check
470 | self._check_files(
471 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 996, in _check_files
472 | self._check_file(get_ast, check_astroid_module, name, filepath, modname)
473 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1022, in _check_file
474 | check_astroid_module(ast_node)
475 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1159, in check_astroid_module
476 | retval = self._check_astroid_module(
477 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1204, in _check_astroid_module
478 | walker.walk(ast_node)
479 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
480 | self.walk(child)
481 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
482 | self.walk(child)
483 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
484 | self.walk(child)
485 | [Previous line repeated 2 more times]
486 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 74, in walk
487 | callback(astroid)
488 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 885, in visit_call
489 | self._check_consider_using_with(node)
490 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1302, in _check_consider_using_with
491 | if could_be_used_in_with and not _is_inside_context_manager(node):
492 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 112, in _is_inside_context_manager
493 | return frame.name == "__enter__" or utils.decorated_with(
494 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/checkers/utils.py", line 846, in decorated_with
495 | if any(
496 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-uUHobIbO/lib/python3.8/site-packages/pylint/checkers/utils.py", line 847, in <genexpr>
497 | i is not None and i.qname() in qnames or i.name in qnames
498 | TypeError: 'in <string>' requires string as left operand, not Uninferable
499 | make: *** [Makefile:33: lint] Error 1
```

If I comment out everything but the first fixture in the test file it now moves on and fails on the next test file:

```
Exception on node <Call l.66 at 0x7fa704e89100> in file '/codebuild/output/src534352411/src//ind-openit-export-forecasting-service/src/test/unit/export_forecasting_controller_tests.py'
568 | Traceback (most recent call last):
569 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/bin/pylint", line 8, in <module>
570 | sys.exit(run_pylint())
571 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/__init__.py", line 24, in run_pylint
572 | PylintRun(sys.argv[1:])
573 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/run.py", line 384, in __init__
574 | linter.check(args)
575 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 962, in check
576 | self._check_files(
577 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 996, in _check_files
578 | self._check_file(get_ast, check_astroid_module, name, filepath, modname)
579 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1022, in _check_file
580 | check_astroid_module(ast_node)
581 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1159, in check_astroid_module
582 | retval = self._check_astroid_module(
583 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1204, in _check_astroid_module
584 | walker.walk(ast_node)
585 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
586 | self.walk(child)
587 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
588 | self.walk(child)
589 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
590 | self.walk(child)
591 | [Previous line repeated 2 more times]
592 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 74, in walk
593 | callback(astroid)
594 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 885, in visit_call
595 | self._check_consider_using_with(node)
596 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1302, in _check_consider_using_with
597 | if could_be_used_in_with and not _is_inside_context_manager(node):
598 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 112, in _is_inside_context_manager
599 | return frame.name == "__enter__" or utils.decorated_with(
600 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/checkers/utils.py", line 846, in decorated_with
601 | if any(
602 | File "/root/.local/share/virtualenvs/ind-openit-export-forecasting-service-T348z7Ej/lib/python3.8/site-packages/pylint/checkers/utils.py", line 847, in <genexpr>
603 | i is not None and i.qname() in qnames or i.name in qnames
604 | TypeError: 'in <string>' requires string as left operand, not Uninferable
605 | make: *** [Makefile:33: lint] Error 1
606


```

--- a/ChangeLog
+++ b/ChangeLog
@@ -14,6 +14,11 @@ Release date: TBA
 * ``setuptools_scm`` has been removed and replaced by ``tbump`` in order to not
   have hidden runtime dependencies to setuptools
 
+* Fix a crash when a test function is decorated with ``@pytest.fixture`` and astroid can't
+  infer the name of the decorator when using ``open`` without ``with``.
+
+  Closes #4612
+
 * Appveyor is no longer used in the continuous integration
 
 * Added ``deprecated-decorator``: Emitted when deprecated decorator is used.

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -810,8 +810,9 @@ def decorated_with(
             decorator_node = decorator_node.func
         try:
             if any(
-                i is not None and i.qname() in qnames or i.name in qnames
+                i.name in qnames or i.qname() in qnames
                 for i in decorator_node.infer()
+                if i is not None and i != astroid.Uninferable
             ):
                 return True
         except astroid.InferenceError:

--- a/tests/functional/r/regression/regression_4612_crash_pytest_fixture.py
+++ b/tests/functional/r/regression/regression_4612_crash_pytest_fixture.py
@@ -1,0 +1,9 @@
+# pylint: disable=missing-docstring,consider-using-with,redefined-outer-name
+
+import pytest
+
+
+@pytest.fixture
+def qm_file():
+    qm_file = open("src/test/resources/example_qm_file.csv").read()
+    return qm_file

